File: src/bin/ingest_sample.rs
//! Ingest a sample of vectors and report per-vector dimension variance.
//!
//! Pre-ingestion variance is computed on raw dataset vectors.
//! Post-ingestion variance is computed on L2-normalized + SRHT-rotated vectors.
//!
//! Usage:
//! ```text
//! cargo run --release --bin ingest-sample -- --npy data/.cache/glove-sample-100.npy
//! cargo run --release --features glove --bin ingest-sample -- --dataset glove --limit 100
//! ```

use std::path::PathBuf;
use std::time::Instant;

use anyhow::{bail, Context, Result};
use clap::Parser;
use vectorcache::datasets::npy::NpyReader;
use vectorcache::datasets::reader::{DatasetReader, DatasetSplit};
#[cfg(feature = "glove")]
use vectorcache::datasets::reader::open_dataset;
use vectorcache::datasets::DatasetKind;
use vectorcache::ingest::{IngestionEngine, TimingSummary, VectorHook};
use vectorcache::transform::{l2_normalize_in_place, padded_dim, SrhtRotation};

#[derive(Debug, Parser)]
#[command(
    name = "ingest-sample",
    about = "Ingest a limited number of vectors and report per-vector dimension variance"
)]
struct Args {
    /// Pre-extracted float32 NPY matrix (bypasses dataset reader)
    #[arg(long)]
    npy: Option<PathBuf>,

    /// Dataset name (default: VECTORCACHE_DATASET env or glove)
    #[arg(long, env = "VECTORCACHE_DATASET", conflicts_with = "npy")]
    dataset: Option<String>,

    /// Data directory (default: VECTORCACHE_DATA_DIR env or data)
    #[arg(long, env = "VECTORCACHE_DATA_DIR", default_value = "data")]
    data_dir: PathBuf,

    /// Maximum number of vectors to ingest
    #[arg(long, default_value_t = 100)]
    limit: usize,

    /// HDF5 split for GloVe (train or test)
    #[arg(long, default_value = "train")]
    split: String,

    /// SRHT rotation seed (each round uses seed + round index)
    #[arg(long, default_value_t = 42)]
    seed: u64,

    /// Number of consecutive SRHT rounds to apply during ingestion
    #[arg(long, default_value_t = 1)]
    rounds: usize,

    /// Print the stored vector at this index after ingestion
    #[arg(long)]
    show_index: Option<usize>,
}

enum SampleReader {
    Npy(NpyReader),
    #[cfg(feature = "glove")]
    Open(vectorcache::datasets::reader::OpenDatasetReader),
}

impl DatasetReader for SampleReader {
    fn meta(&self) -> vectorcache::datasets::reader::DatasetMeta {
        match self {
            Self::Npy(r) => r.meta(),
            #[cfg(feature = "glove")]
            Self::Open(r) => r.meta(),
        }
    }

    fn next_vector_into(&mut self, out: &mut [f32]) -> Result<bool> {
        match self {
            Self::Npy(r) => r.next_vector_into(out),
            #[cfg(feature = "glove")]
            Self::Open(r) => r.next_vector_into(out),
        }
    }
}

struct LimitedReader<R> {
    inner: R,
    remaining: usize,
    pre_variances: Vec<f64>,
}

impl<R: DatasetReader> DatasetReader for LimitedReader<R> {
    fn meta(&self) -> vectorcache::datasets::reader::DatasetMeta {
        self.inner.meta()
    }

    fn next_vector_into(&mut self, out: &mut [f32]) -> Result<bool> {
        if self.remaining == 0 {
            return Ok(false);
        }

        if !self
            .inner
            .next_vector_into(out)?
        {
            bail!("reader exhausted before reaching ingest limit");
        }
        self.pre_variances.push(variance_across_dims(out));
        self.remaining -= 1;
        Ok(true)
    }
}

struct VarianceHook {
    post_variances: Vec<f64>,
    vectors: Vec<Vec<f32>>,
    capture_vectors: bool,
}

impl VarianceHook {
    fn new(capture_vectors: bool) -> Self {
        Self {
            post_variances: Vec::new(),
            vectors: Vec::new(),
            capture_vectors,
        }
    }
}

impl VectorHook for VarianceHook {
    fn on_vector(&mut self, _global_id: u64, vector: &[f32]) -> Result<()> {
        self.post_variances.push(variance_across_dims(vector));
        if self.capture_vectors {
            self.vectors.push(vector.to_vec());
        }
        Ok(())
    }
}

struct MultiRoundReader<R> {
    inner: R,
    remaining: usize,
    pre_variances: Vec<f64>,
    rounds: usize,
    seed: u64,
    round_variances: Vec<Vec<f64>>,
    raw: Vec<f32>,
    scratch: Vec<f32>,
    output: Vec<f32>,
}

impl<R: DatasetReader> DatasetReader for MultiRoundReader<R> {
    fn meta(&self) -> vectorcache::datasets::reader::DatasetMeta {
        self.inner.meta()
    }

    fn next_vector_into(&mut self, out: &mut [f32]) -> Result<bool> {
        if self.remaining == 0 {
            return Ok(false);
        }

        if !self.inner.next_vector_into(&mut self.raw)? {
            bail!("reader exhausted before reaching ingest limit");
        }
        self.pre_variances.push(variance_across_dims(&self.raw));
        self.remaining -= 1;

        self.output.copy_from_slice(&self.raw);
        l2_normalize_in_place(&mut self.output);
        for round in 0..self.rounds {
            let rot = SrhtRotation::new(self.output.len(), self.seed + round as u64);
            self.scratch.resize(rot.padded_dim(), 0.0);
            rot.apply(&self.output, &mut self.scratch)?;
            if self.round_variances.len() <= round {
                self.round_variances.push(Vec::new());
            }
            self.round_variances[round].push(variance_across_dims(&self.scratch));
            self.output.copy_from_slice(&self.scratch);
        }

        if out.len() != self.output.len() {
            bail!(
                "output buffer dimension mismatch: expected {}, got {}",
                self.output.len(),
                out.len()
            );
        }
        out.copy_from_slice(&self.output);
        Ok(true)
    }
}

fn main() -> Result<()> {
    let args = Args::parse();
    let split = match args.split.as_str() {
        "train" => DatasetSplit::Train,
        "test" => DatasetSplit::Test,
        other => bail!("unknown split '{other}'; use train or test"),
    };

    if args.rounds == 0 {
        bail!("--rounds must be at least 1");
    }

    let reader = if let Some(path) = &args.npy {
        SampleReader::Npy(NpyReader::open(path, "glove-sample")?)
    } else {
        let name = args.dataset.as_deref().unwrap_or("glove");
        let kind = DatasetKind::parse(name)
            .with_context(|| format!("unknown dataset '{name}'"))?;
        #[cfg(feature = "glove")]
        {
            SampleReader::Open(open_dataset(kind, &args.data_dir, split)?)
        }
        #[cfg(not(feature = "glove"))]
        {
            let _ = (kind, split);
            bail!(
                "dataset '{name}' requires the 'glove' feature or pass --npy with a pre-extracted sample",
            );
        }
    };

    let meta = reader.meta();
    let limit = args.limit.min(meta.count);
    let padded = padded_dim(meta.dim);
    let capture_vectors = args.show_index.is_some();

    println!(
        "Dataset: {} (dim={}, padded={}, available={}, ingesting={}, srht_seed={}, rounds={})",
        meta.label, meta.dim, padded, meta.count, limit, args.seed, args.rounds
    );

    let mut hook = VarianceHook::new(capture_vectors);
    hook.post_variances.reserve(limit);
    if capture_vectors {
        hook.vectors.reserve(limit);
    }

    let ingest_start = Instant::now();

    let (pre_variances, round_variances, report, engine) = if args.rounds == 1 {
        let mut limited = LimitedReader {
            inner: reader,
            remaining: limit,
            pre_variances: Vec::with_capacity(limit),
        };
        let mut engine = IngestionEngine::with_rotation(meta.dim, args.seed);
        engine.reserve_vectors(limit);
        let mut hook_ref = Some(&mut hook);
        let report = engine.ingest(&mut limited, &mut hook_ref)?;
        (limited.pre_variances, Vec::new(), report, engine)
    } else {
        let mut limited = MultiRoundReader {
            inner: reader,
            remaining: limit,
            pre_variances: Vec::with_capacity(limit),
            rounds: args.rounds,
            seed: args.seed,
            round_variances: Vec::new(),
            raw: vec![0.0; meta.dim],
            scratch: Vec::new(),
            output: vec![0.0; meta.dim],
        };
        let store_dim = {
            let mut dim = meta.dim;
            for _ in 0..args.rounds {
                dim = padded_dim(dim);
            }
            dim
        };
        let mut engine = IngestionEngine::from_rotated(store_dim);
        engine.reserve_vectors(limit);
        let mut hook_ref = Some(&mut hook);
        let report = engine.ingest(&mut limited, &mut hook_ref)?;
        (
            limited.pre_variances,
            limited.round_variances,
            report,
            engine,
        )
    };

    let elapsed_ns = ingest_start.elapsed().as_nanos() as u64;

    print_variance_stats("Pre-ingestion (raw)", &pre_variances);
    for (i, round_stats) in round_variances.iter().enumerate() {
        print_variance_stats(&format!("After SRHT round {}", i + 1), round_stats);
    }
    print_variance_stats("Post-ingestion (stored)", &hook.post_variances);

    println!("Ingested: {} vectors", report.vectors_ingested);
    println!(
        "Total ingestion time: {} ({elapsed_ns:.0} ns)",
        TimingSummary::format_duration(elapsed_ns)
    );
    if report.vectors_ingested > 0 {
        let per_vec = elapsed_ns as f64 / report.vectors_ingested as f64;
        println!(
            "  per-vector: {} ({per_vec:.0} ns)",
            TimingSummary::format_duration(per_vec as u64)
        );
    }
    println!(
        "Blocks: {} full, {} in partial block (L1 words/vec: {})",
        report.full_blocks,
        report.partial_len,
        engine.store().l1_words_per_vec()
    );

    if let Some(index) = args.show_index {
        if index >= report.vectors_ingested as usize {
            bail!(
                "--show-index {index} out of range (ingested {})",
                report.vectors_ingested
            );
        }
        print_stored_l1_codes(index, &engine, padded);
        if let Some(vector) = hook.vectors.get(index) {
            print_rotated_vector(index, vector);
        }
    }

    Ok(())
}

fn print_stored_l1_codes(index: usize, engine: &IngestionEngine, padded_dim: usize) {
    let words_per_vec = engine.store().l1_words_per_vec();
    let num_bits = (padded_dim + 3) / 4;
    let block_idx = index / vectorcache::ingest::BLOCK_SIZE;
    let vec_in_block = index % vectorcache::ingest::BLOCK_SIZE;

    let block = if block_idx < engine.store().block_count() {
        engine.store().get_block(block_idx).unwrap()
    } else {
        engine.store().partial_block()
    };

    let offset = vec_in_block * words_per_vec;
    let codes = &block.as_slice()[offset..offset + words_per_vec];
    let hex: Vec<String> = codes.iter().map(|w| format!("{w:#018x}")).collect();
    println!(
        "Stored L1 codes at index {index} ({num_bits} bits, {} u64 words):",
        words_per_vec
    );
    println!("  [{}]", hex.join(", "));
}

fn print_rotated_vector(index: usize, vector: &[f32]) {
    println!(
        "Rotated f32 at index {index} (dim={}, L2 norm={:.6}):",
        vector.len(),
        l2_norm(vector)
    );
    let preview = 16.min(vector.len());
    let head: Vec<String> = vector[..preview]
        .iter()
        .map(|x| format!("{x:+.6}"))
        .collect();
    println!("  [{}]", head.join(", "));
    if vector.len() > preview {
        println!("  ... ({} more dims)", vector.len() - preview);
    }
}

fn l2_norm(vector: &[f32]) -> f64 {
    vector.iter().map(|&x| x as f64 * x as f64).sum::<f64>().sqrt()
}

fn print_variance_stats(label: &str, variances: &[f64]) {
    if variances.is_empty() {
        println!("{label}: (no vectors)");
        return;
    }
    let avg = variances.iter().sum::<f64>() / variances.len() as f64;
    let min = variances.iter().copied().fold(f64::INFINITY, f64::min);
    let max = variances.iter().copied().fold(f64::NEG_INFINITY, f64::max);

    println!("{label} per-vector variance across dims:");
    println!("  average: {avg:.8}");
    println!("  min:     {min:.8}");
    println!("  max:     {max:.8}");
}

/// Population variance of a single vector's components across its dimensions.
fn variance_across_dims(vector: &[f32]) -> f64 {
    if vector.is_empty() {
        return 0.0;
    }

    let n = vector.len() as f64;
    let mean = vector.iter().map(|&x| x as f64).sum::<f64>() / n;
    let sum_sq = vector
        .iter()
        .map(|&x| {
            let d = x as f64 - mean;
            d * d
        })
        .sum::<f64>();
    sum_sq / n
}

File: tools/ingest-sample.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <limits>
#include <memory>
#include <optional>
#include <sstream>
#include <string>
#include <vector>

#include <CLI/CLI.hpp>

#include "vectorcache/datasets/datasets.hpp"
#include "vectorcache/datasets/npy.hpp"
#include "vectorcache/error.hpp"
#include "vectorcache/ingest/engine.hpp"
#include "vectorcache/ingest/hook.hpp"
#include "vectorcache/ingest/timing.hpp"
#include "vectorcache/transform/fwht.hpp"
#include "vectorcache/transform/normalize.hpp"
#include "vectorcache/transform/srht.hpp"

namespace {

double variance_across_dims(std::span<const float> vector) {
  if (vector.empty()) return 0.0;
  const double n = static_cast<double>(vector.size());
  double mean = 0.0;
  for (float x : vector) mean += static_cast<double>(x);
  mean /= n;
  double sum_sq = 0.0;
  for (float x : vector) {
    const double d = static_cast<double>(x) - mean;
    sum_sq += d * d;
  }
  return sum_sq / n;
}

void print_variance_stats(const std::string& label, const std::vector<double>& variances) {
  if (variances.empty()) {
    std::cout << label << ": (no vectors)\n";
    return;
  }
  double avg = 0.0;
  double min_v = std::numeric_limits<double>::infinity();
  double max_v = -std::numeric_limits<double>::infinity();
  for (double v : variances) {
    avg += v;
    min_v = std::min(min_v, v);
    max_v = std::max(max_v, v);
  }
  avg /= static_cast<double>(variances.size());
  std::cout << label << " per-vector variance across dims:\n";
  std::cout << "  average: " << std::fixed << std::setprecision(8) << avg << '\n';
  std::cout << "  min:     " << min_v << '\n';
  std::cout << "  max:     " << max_v << '\n';
}

class VarianceHook : public vectorcache::ingest::VectorHook {
 public:
  explicit VarianceHook(bool capture_vectors) : capture_vectors_(capture_vectors) {}

  void on_vector(std::uint64_t /*global_id*/, std::span<const float> vector) override {
    post_variances_.push_back(variance_across_dims(vector));
    if (capture_vectors_) {
      vectors_.emplace_back(vector.begin(), vector.end());
    }
  }

  const std::vector<double>& post_variances() const { return post_variances_; }
  const std::vector<std::vector<float>>& vectors() const { return vectors_; }

 private:
  bool capture_vectors_;
  std::vector<double> post_variances_;
  std::vector<std::vector<float>> vectors_;
};

class LimitedReader : public vectorcache::datasets::DatasetReader {
 public:
  LimitedReader(vectorcache::datasets::DatasetReader& inner, std::size_t remaining,
                std::vector<double>* pre_variances)
      : inner_(inner), remaining_(remaining), pre_variances_(pre_variances) {}

  vectorcache::datasets::DatasetMeta meta() const override { return inner_.meta(); }

  bool next_vector_into(std::span<float> out) override {
    if (remaining_ == 0) return false;
    if (!inner_.next_vector_into(out)) {
      throw vectorcache::Error("reader exhausted before reaching ingest limit");
    }
    if (pre_variances_ != nullptr) {
      pre_variances_->push_back(variance_across_dims(out));
    }
    --remaining_;
    return true;
  }

 private:
  vectorcache::datasets::DatasetReader& inner_;
  std::size_t remaining_;
  std::vector<double>* pre_variances_;
};

class MultiRoundReader : public vectorcache::datasets::DatasetReader {
 public:
  MultiRoundReader(vectorcache::datasets::DatasetReader& inner, std::size_t remaining,
                   std::vector<double>* pre_variances, std::size_t rounds, std::uint64_t seed,
                   std::vector<std::vector<double>>* round_variances)
      : inner_(inner),
        remaining_(remaining),
        pre_variances_(pre_variances),
        rounds_(rounds),
        seed_(seed),
        round_variances_(round_variances) {
    raw_.resize(inner.meta().dim);
    output_.resize(inner.meta().dim);
  }

  vectorcache::datasets::DatasetMeta meta() const override { return inner_.meta(); }

  bool next_vector_into(std::span<float> out) override {
    if (remaining_ == 0) return false;
    if (!inner_.next_vector_into(raw_)) {
      throw vectorcache::Error("reader exhausted before reaching ingest limit");
    }
    if (pre_variances_ != nullptr) {
      pre_variances_->push_back(variance_across_dims(raw_));
    }
    --remaining_;

    output_ = raw_;
    vectorcache::transform::l2_normalize_in_place(output_);
    for (std::size_t round = 0; round < rounds_; ++round) {
      vectorcache::transform::SrhtRotation rot(output_.size(), seed_ + round);
      scratch_.assign(rot.padded_dim(), 0.0f);
      rot.apply(output_, scratch_);
      if (round_variances_ != nullptr && !round_variances_->empty()) {
        (*round_variances_)[round].push_back(variance_across_dims(scratch_));
      }
      output_ = scratch_;
    }

    if (out.size() != output_.size()) {
      throw vectorcache::Error("output buffer dimension mismatch");
    }
    std::copy(output_.begin(), output_.end(), out.begin());
    return true;
  }

 private:
  vectorcache::datasets::DatasetReader& inner_;
  std::size_t remaining_;
  std::vector<double>* pre_variances_;
  std::size_t rounds_;
  std::uint64_t seed_;
  std::vector<std::vector<double>>* round_variances_;
  std::vector<float> raw_;
  std::vector<float> scratch_;
  std::vector<float> output_;
};

void print_stored_l1_codes(std::size_t index, const vectorcache::ingest::IngestionEngine& engine,
                           std::size_t padded_dim) {
  const std::size_t words_per_vec = engine.store().l1_words_per_vec();
  const std::size_t num_bits = (padded_dim + 3) / 4;
  const std::size_t block_idx = index / vectorcache::ingest::BLOCK_SIZE;
  const std::size_t vec_in_block = index % vectorcache::ingest::BLOCK_SIZE;

  const vectorcache::ingest::VectorBlock* block = nullptr;
  if (block_idx < engine.store().block_count()) {
    block = engine.store().get_block(block_idx);
  } else {
    block = &engine.store().partial_block();
  }

  const auto slice = block->as_slice();
  const std::size_t offset = vec_in_block * words_per_vec;
  std::cout << "Stored L1 codes at index " << index << " (" << num_bits << " bits, "
            << words_per_vec << " u64 words):\n  [";
  for (std::size_t i = 0; i < words_per_vec; ++i) {
    if (i > 0) std::cout << ", ";
    std::cout << "0x" << std::hex << std::setw(16) << std::setfill('0') << slice[offset + i];
  }
  std::cout << std::dec << "]\n";
}

}  // namespace

int main(int argc, char** argv) {
  CLI::App app{"Ingest vectors and optionally report per-vector dimension variance"};
  std::filesystem::path npy_path;
  std::string dataset;
  std::filesystem::path data_dir = "data";
  std::size_t limit = 100;
  std::string split = "train";
  std::uint64_t seed = 42;
  std::size_t rounds = 1;
  bool variance = false;
  std::optional<std::size_t> show_index;

  app.add_option("--npy", npy_path, "Pre-extracted float32 NPY matrix");
  app.add_option("--dataset", dataset, "Dataset name")->envname("VECTORCACHE_DATASET");
  app.add_option("--data-dir", data_dir, "Data directory")->envname("VECTORCACHE_DATA_DIR");
  app.add_option("--limit", limit, "Maximum number of vectors to ingest");
  app.add_option("--split", split, "HDF5 split for GloVe (train or test)");
  app.add_option("--seed", seed, "SRHT rotation seed");
  app.add_option("--rounds", rounds, "Number of consecutive SRHT rounds");
  app.add_flag("--variance", variance, "Report per-vector dimension variance");
  app.add_option("--show-index", show_index, "Print stored vector at index");

  CLI11_PARSE(app, argc, argv);

  try {
    if (rounds == 0) {
      throw vectorcache::Error("--rounds must be at least 1");
    }

    vectorcache::datasets::DatasetSplit dataset_split = vectorcache::datasets::DatasetSplit::Train;
    if (split == "test") {
      dataset_split = vectorcache::datasets::DatasetSplit::Test;
    } else if (split != "train") {
      throw vectorcache::Error("unknown split '" + split + "'; use train or test");
    }

    std::unique_ptr<vectorcache::datasets::DatasetReader> owned_reader;
    vectorcache::datasets::DatasetReader* reader_ptr = nullptr;

    if (!npy_path.empty()) {
      owned_reader =
          std::make_unique<vectorcache::datasets::NpyReader>(
              vectorcache::datasets::NpyReader::open(npy_path, "glove-sample"));
      reader_ptr = owned_reader.get();
    } else {
      const std::string name = dataset.empty() ? "glove" : dataset;
      auto kind = vectorcache::datasets::parse_dataset_kind(name);
      if (!kind) {
        throw vectorcache::Error("unknown dataset '" + name + "'");
      }
      owned_reader = std::make_unique<vectorcache::datasets::OpenDatasetReader>(
          vectorcache::datasets::OpenDatasetReader::open(*kind, data_dir, dataset_split));
      reader_ptr = owned_reader.get();
    }

    const auto meta = reader_ptr->meta();
    const std::size_t ingest_limit = std::min(limit, meta.count);
    const std::size_t padded = vectorcache::transform::padded_dim(meta.dim);
    const bool capture_vectors = show_index.has_value() && variance;

    std::cout << "Dataset: " << meta.label << " (dim=" << meta.dim << ", padded=" << padded
              << ", available=" << meta.count << ", ingesting=" << ingest_limit
              << ", srht_seed=" << seed << ", rounds=" << rounds << ")\n";

    VarianceHook variance_hook(capture_vectors);
    const auto ingest_start = std::chrono::steady_clock::now();

    std::vector<double> pre_variances;
    std::vector<std::vector<double>> round_variances;
    vectorcache::ingest::IngestReport report{};
    vectorcache::ingest::IngestionEngine engine =
        vectorcache::ingest::IngestionEngine::with_rotation(meta.dim, seed);

    if (rounds == 1) {
      LimitedReader limited(*reader_ptr, ingest_limit, variance ? &pre_variances : nullptr);
      engine.reserve_vectors(ingest_limit);
      if (variance) {
        report = engine.ingest_with_hook(limited, &variance_hook);
      } else {
        report = engine.ingest(limited);
      }
    } else {
      if (variance) {
        round_variances.resize(rounds);
        for (auto& rv : round_variances) rv.reserve(ingest_limit);
      }
      MultiRoundReader limited(*reader_ptr, ingest_limit, variance ? &pre_variances : nullptr,
                               rounds, seed, variance ? &round_variances : nullptr);
      std::size_t store_dim = meta.dim;
      for (std::size_t r = 0; r < rounds; ++r) {
        store_dim = vectorcache::transform::padded_dim(store_dim);
      }
      engine = vectorcache::ingest::IngestionEngine::from_rotated(store_dim);
      engine.reserve_vectors(ingest_limit);
      if (variance) {
        report = engine.ingest_with_hook(limited, &variance_hook);
      } else {
        report = engine.ingest(limited);
      }
    }

    const auto elapsed_ns = static_cast<std::uint64_t>(
        std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                             ingest_start)
            .count());

    if (variance) {
      print_variance_stats("Pre-ingestion (raw)", pre_variances);
      for (std::size_t i = 0; i < round_variances.size(); ++i) {
        print_variance_stats("After SRHT round " + std::to_string(i + 1), round_variances[i]);
      }
      print_variance_stats("Post-ingestion (stored)", variance_hook.post_variances());
    }

    std::cout << "Ingested: " << report.vectors_ingested << " vectors\n";
    std::cout << "Total ingestion time: "
              << vectorcache::ingest::TimingSummary::format_duration(elapsed_ns) << " ("
              << elapsed_ns << " ns)\n";
    if (report.vectors_ingested > 0) {
      const double per_vec = static_cast<double>(elapsed_ns) / report.vectors_ingested;
      std::cout << "  per-vector: "
                << vectorcache::ingest::TimingSummary::format_duration(
                       static_cast<std::uint64_t>(per_vec))
                << " (" << per_vec << " ns)\n";
    }
    std::cout << "Blocks: " << report.full_blocks << " full, " << report.partial_len
              << " in partial block (L1 words/vec: " << engine.store().l1_words_per_vec()
              << ")\n";

    if (show_index) {
      if (*show_index >= report.vectors_ingested) {
        throw vectorcache::Error("--show-index out of range");
      }
      print_stored_l1_codes(*show_index, engine, padded);
      const auto& vectors = variance_hook.vectors();
      if (*show_index < vectors.size()) {
        const auto& vector = vectors[*show_index];
        double norm = 0.0;
        for (float x : vector) norm += static_cast<double>(x) * x;
        norm = std::sqrt(norm);
        std::cout << "Rotated f32 at index " << *show_index << " (dim=" << vector.size()
                  << ", L2 norm=" << std::fixed << std::setprecision(6) << norm << "):\n";
        const std::size_t preview = std::min<std::size_t>(16, vector.size());
        std::cout << "  [";
        for (std::size_t i = 0; i < preview; ++i) {
          if (i > 0) std::cout << ", ";
          std::cout << std::showpos << std::fixed << std::setprecision(6) << vector[i]
                    << std::noshowpos;
        }
        std::cout << "]\n";
        if (vector.size() > preview) {
          std::cout << "  ... (" << (vector.size() - preview) << " more dims)\n";
        }
      } else if (!variance) {
        std::cout << "(pass --variance with --show-index to also print the rotated f32 vector)\n";
      }
    }

    return 0;
  } catch (const vectorcache::Error& e) {
    std::cerr << e.what() << '\n';
    return 1;
  }
}
